ext/opcache/Optimizer/dce.c

```diff
@@ -0,0 +1,647 @@
+ * that all instructions and phis are dead. Instructions with immediate side-effects are then marked
+ * as live. We then recursively (using a worklist) propagate liveness to the instructions that def
+ * the used operands.
+ *
+ * Notes:
+ *  * This pass does not perform unreachable code elimination. This happens as part of the SCCP
+ *    pass.
+ *  * The DCE is performed without taking control-dependence into account, i.e. all conditional
+ *    branches are assumed to be live. It's possible to take control-dependence into account using
+ *    the DCE algorithm described by Cytron et al., however it requires the construction of a
+ *    postdominator tree and of postdominance frontiers, which does not seem worthwhile at this
+ *    point.
+ *  * We separate intrinsic side-effects from potential side-effects in the form of notices thrown
+ *    by the instruction (in case we want to make this configurable). See may_have_side_effect() and
+ *    zend_may_throw().
+ *  * We often cannot DCE assignments and unsets while guaranteeing that dtors run in the same
+ *    order. There is an optimization option to allow reordering of dtor effects.
+ */
+
+typedef struct {
+	zend_ssa *ssa;
+	zend_op_array *op_array;
+	zend_bitset instr_dead;
+	zend_bitset phi_dead;
+	zend_bitset instr_worklist;
+	zend_bitset phi_worklist;
+	zend_bitset phi_worklist_no_val;
+	uint32_t instr_worklist_len;
+	uint32_t phi_worklist_len;
+	unsigned reorder_dtor_effects : 1;
+} context;
+
+static inline zend_bool is_bad_mod(const zend_ssa *ssa, int use, int def) {
+	if (def < 0) {
+		/* This modification is not tracked by SSA, assume the worst */
+		return 1;
+	}
+	if (ssa->var_info[use].type & MAY_BE_REF) {
+		/* Modification of reference may have side-effect */
+		return 1;
+	}
+	return 0;
+}
+
+static inline zend_bool may_have_side_effects(
+		zend_op_array *op_array, zend_ssa *ssa,
+		const zend_op *opline, const zend_ssa_op *ssa_op,
+		zend_bool reorder_dtor_effects) {
+	switch (opline->opcode) {
+		case ZEND_NOP:
+		case ZEND_IS_IDENTICAL:
+		case ZEND_IS_NOT_IDENTICAL:
+		case ZEND_QM_ASSIGN:
+		case ZEND_FREE:
+		case ZEND_TYPE_CHECK:
+		case ZEND_DEFINED:
+		case ZEND_ADD:
+		case ZEND_SUB:
+		case ZEND_MUL:
+		case ZEND_POW:
+		case ZEND_BW_OR:
+		case ZEND_BW_AND:
+		case ZEND_BW_XOR:
+		case ZEND_CONCAT:
+		case ZEND_FAST_CONCAT:
+		case ZEND_DIV:
+		case ZEND_MOD:
+		case ZEND_BOOL_XOR:
+		case ZEND_BOOL:
+		case ZEND_BOOL_NOT:
+		case ZEND_BW_NOT:
+		case ZEND_SL:
+		case ZEND_SR:
+		case ZEND_IS_EQUAL:
+		case ZEND_IS_NOT_EQUAL:
+		case ZEND_IS_SMALLER:
+		case ZEND_IS_SMALLER_OR_EQUAL:
+		case ZEND_CASE:
+		case ZEND_CAST:
+		case ZEND_ROPE_INIT:
+		case ZEND_ROPE_ADD:
+		case ZEND_ROPE_END:
+		case ZEND_INIT_ARRAY:
+		case ZEND_ADD_ARRAY_ELEMENT:
+		case ZEND_SPACESHIP:
+		case ZEND_STRLEN:
+		case ZEND_COUNT:
+		case ZEND_GET_TYPE:
+		case ZEND_ISSET_ISEMPTY_THIS:
+		case ZEND_ISSET_ISEMPTY_DIM_OBJ:
+		case ZEND_FETCH_DIM_IS:
```

Comment:
`zend_may_throw()` for FETCH_DIM_IS checks for `(t1 & MAY_BE_OBJECT) || (t2 & (MAY_BE_ARRAY|MAY_BE_OBJECT)`. After checking the code a resource as t2 also throws a notice. However, I think that this is an engine bug:

```php
<?php
$arr = [];
isset($arr[STDOUT]); // silent
isset($arr[STDOUT][0]); // Notice: Resource ID#2 used as offset, casting to integer (2)
```

These two should behave the same way...